wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileManager.java

```diff
@@ -121,23 +187,85 @@ public void resolveReference(DevfileImpl devfile, FileContentProvider fileConten
+      throw new DevfileFormatException("Unable to parse Devfile. Error: " + e.getMessage());
     }
-
     integrityValidator.validateDevfile(devfile);
     return devfile;
   }
 
-  @FunctionalInterface
-  private interface ValidationFunction {
-    JsonNode validate(String content) throws DevfileFormatException;
+  private JsonNode applyPropertiesOverride(
+      JsonNode devfileNode, Map<String, String> overrideProperties) throws DevfileFormatException {
+    for (Map.Entry<String, String> entry : overrideProperties.entrySet()) {
+      // prepare stuff
+      String[] pathSegments = entry.getKey().split("\\.");
+      if (pathSegments.length < 1) {
+        continue;
+      }
+      String lastSegment = pathSegments[pathSegments.length - 1];
+      JsonNode currentNode = devfileNode;
+      // iterate until we reach last but one path segment,
+      // because it's impossible to change value of the current Json node,
+      // so to set value to Json node, you should be 1 level up and do put(key, value).
+      Iterator<String> pathSegmentsIterator =
+          asList(copyOf(pathSegments, pathSegments.length)).iterator();
+      do {
+        String currentSegment = pathSegmentsIterator.next();
+        JsonNode nextNode = currentNode.path(currentSegment);
+        if (nextNode.isMissingNode() && pathSegmentsIterator.hasNext()) {
+          // no such intermediate node, let's create it as a empty object
+          currentNode = ((ObjectNode) currentNode).putObject(currentSegment);
+          continue;
+        } else if (nextNode.isArray()) {
+          // ok we have reference to array, so need to make sure that we have next path segment
+          // ant then try to retrieve it from array
+          if (!pathSegmentsIterator.hasNext()) {
+            throw new DevfileFormatException(
+                format(
+                    "Override property reference '%s' points to an array type object. Please add an item qualifier by name or alias.",
+                    entry.getKey()));
+          }
+          // retrieve object by name from array
+          String arrayObjectName = pathSegmentsIterator.next();
+          Optional<JsonNode> namedNode =
+              findNodeByNameOrAlias((ArrayNode) nextNode, arrayObjectName);
+          currentNode =
+              namedNode.orElseThrow(
+                  () ->
+                      new DevfileFormatException(
+                          format(
+                              "Object with name '%s' not found in array of %s.",
+                              arrayObjectName, currentSegment)));
+          continue;
+        } else {
+          if (pathSegmentsIterator.hasNext()) {
+            // do not set last segment as current node, we need to be 1 level up
+            currentNode = nextNode;
+          }
+        }
+      } while (pathSegmentsIterator.hasNext());
+      // end of path segments reached, now we can set value
+      ((ObjectNode) currentNode).put(lastSegment, entry.getValue());
```

Comment:
I haven't tested it but what happens if the last segment refers to child in an array?

Reply:
So child will be replaced with an value (not sure if it's very consistent, but what we can do better?)